### PR TITLE
Implementation of variance_highbd_avx2 kernel

### PR DESCRIPTION
**Pull request by tszumski**
For 1080p 10-bit video in "preset 8" ~2.5% speedup

### CODE REVIEW

**Comment by joelsole** [comment]
It might be worth adding an assert for the supported block sizes, as they are not the same as the c implementation.

**Comment by tszumski** [comment]
Hi, 
There are asserts in code for unsupported block sizes in function variance_highbd_avx2
`assert(w == h);`
and 
`default: assert(0);`

**Comment by EricLamVCD** [comment]
Tested obj-1 10bit clips: 
  | Speed Deviation  | BD-Rate   Deviation
-- | -- | --
M0 - 10bit | 0.98% | 0.00%
M4 - 10bit | 0.58% | 0.00%
M8 - 10bit | 1.30% | 0.00%

There is a ~1.30% speed up for M8 10 bit.

**Comment by tszumski** [comment]
Rebased to master (to fix merge conflict)

File: Source/Lib/Encoder/ASM_AVX2/highbd_variance_avx2.c [hunk 1]
```diff
@@ -138,3 +138,38 @@ VAR_FN(16, 64, 16, 10);
 VAR_FN(64, 16, 16, 10);
 
 #undef VAR_FN
+
+void variance_highbd_32x32_avx2(const uint16_t *src, int src_stride, const uint16_t *ref,
```
Comment (fouedb) [inline Source/Lib/Encoder/ASM_AVX2/highbd_variance_avx2.c:142]: Add a description of the function.

File: Source/Lib/Encoder/ASM_AVX2/highbd_variance_avx2.c [hunk 2]
```diff
@@ -138,3 +138,38 @@ VAR_FN(16, 64, 16, 10);
...
+                                         ref_stride,
+                                         &sse0,
+                                         &sum0);
+            *sum += sum0;
+            *sse += sse0;
+        }
+    }
+}
+
+uint32_t variance_highbd_avx2(const uint16_t *a, int a_stride, const uint16_t *b, int b_stride,
```
Comment (fouedb) [inline Source/Lib/Encoder/ASM_AVX2/highbd_variance_avx2.c:161]: Add a description of the function.

File: test/HbdVarianceTest.cc [hunk 3]
```diff
@@ -459,4 +459,133 @@ static const HbdSquareVarianceParam HbdSquareVarTestVector[] = {
...
+    }
+
+    ~HbdSquareVarianceNoRoundTest() {
+        eb_aom_free(src_data_);
+        src_data_ = nullptr;
+        eb_aom_free(ref_data_);
+        ref_data_ = nullptr;
+    }
+
+    void run_zero_test(int times) {
```
Comment (fouedb) [inline test/HbdVarianceTest.cc:494]: Add a description of the function.

File: test/HbdVarianceTest.cc [hunk 4]
```diff
@@ -459,4 +459,133 @@ static const HbdSquareVarianceParam HbdSquareVarTestVector[] = {
...
+                      length_,
+                      length_,
+                      &sse_tst);
+            ASSERT_EQ(sse_tst, 0u) << "Expect 0 sse, got: " << sse_tst;
+            ASSERT_EQ(distortion_tst, 0)
+                << "Expect 0 distortion, got: " << distortion_tst;
+        }
+    }
+
+    void run_maximum_test() {
```
Comment (fouedb) [inline test/HbdVarianceTest.cc:515]: Add a description of the function.

**Comment by hassount** [comment]
@tszumski 
can you please address the documentation comments so that we can merge this ?

**Comment by tszumski** [comment]
Hi @hassount, @fouedb 

I added description

**Comment by hassount** [comment]
@tszumski thank you :) @fouedb good to go ?

**Comment by tszumski** [comment]
Hi All,
Sorry for long delay, typo fixed

**Pull request merged**